### ISSUE TITLE
Encrypt and validate session and/or cache data?

### ISSUE THREAD

**Issue by tom--** [yiisoft/yii2#13146]
**Feature idea**

Data is vulnerable to theft and modification while it is at rest in a cache or session store and while it is in transit between PHP and that store. Users can reduce vulnerability with cryptographic validation to detect tampering or with encryption+validation incorporated into session/cache read and write handling.

I think Yii could support for these functions so that they are relatively easy to use. The user should need to do little more than configure it and provide a function that returns a key within a specific context (e.g. session context).

**Comment by dynasource** [comment]
the first section of your post can directly go in the guide

**Comment by samdark** [comment]
Hmm... that sounds like it's out of scope of the framework. Cache and session storage should never be exposed to public. If they are, you have huge problems even if data is encrypted.

**Comment by klimov-paul** [comment]
Agreed with @samdark : session or cache storage should be hidden from any external influence.
Creating some encryption here is just an overkill like encrypting project PHP source code.

**Comment by samdark** [comment]
I'm closing the issue since it's unlikely to get into framework but feel free to discuss it further. Could be re-opened if needed.

**Comment by cebe** [comment]
Still there are applications where you must protect data to a maximum so that even if someone breaks in, it is very hard for them to get to the data. We should at least check whether Yii easily allows extension to implement such methods.

**Comment by dynasource** [comment]
what if a government demands you to encrypt data you store on different servers throughout the world

**Comment by samdark** [comment]
OK. Checking if enough flexibility is there to achieve it sounds fine.

**Comment by tom--** [comment]
> Cache and session storage should never be exposed to public. If they are, you have huge problems even if data is encrypted.

@samdark i don't understand this. how are they exposed to the public?

this is about security-in-depth, i.e. if a hacker gets onto your redis server, they have a hard time understanding the data.

**Comment by tom--** [comment]
It's really just a matter of doing the crypto on the serialized data that's sent to the custom session/cache store.

**Comment by klimov-paul** [comment]
> Still there are applications where you must protect data to a maximum so that even if someone breaks in, it is very hard for them to get to the data. We should at least check whether Yii easily allows extension to implement such methods.

I do not understand this.
Take a regular SQL Database usage: the channel between application and SQL Server is protected at protocol level using username/password pair.
Do you insinst there should be additional(!) security encryption during data-transfering to SQL server?

> this is about security-in-depth, i.e. if a hacker gets onto your redis server, they have a hard time understanding the data.

If hacker gets to the Redis server I can't see how data encryption will help as any ecnryption key can be broken using `brute-force` approach. If server hacked - the data is lost. That is why it is recommended to change user password after web-site is broken even if passwords are encrypted.

**Comment by samdark** [comment]
OK. Let's review points of extension.

## Cache

There's already [serializer](https://github.com/yiisoft/yii2/blob/master/framework/caching/Cache.php#L73) property so it could be easily extended now. Alternatively you can implement custom cache.

## Session

There's [handler property](https://github.com/yiisoft/yii2/blob/master/framework/web/Session.php#L84) which allows achieving what you need. Alternatively, you can implement custom session.

**Comment by cebe** [comment]
> as any ecnryption key can be broken using brute-force approach

This is not true. MD5 and SHA1 are broken to some extend but that is completely unrelated to symmetric and asymmetric encryption.

**Comment by tom--** [comment]
>  Do you insinst there should be additional(!) security encryption during data-transfering to SQL server?

No, I do not insist that. But your logic would mean by extension there's no need to obfuscate passwords, personal data, or credit card stored in a database because "the channel between application and SQL Server is protected at protocol level" and so on.

> If hacker gets to the Redis server I can't see how data encryption will help as any ecnryption key can be broken using brute-force approach.

No it cannot.

**Comment by klimov-paul** [comment]
> if a hacker gets onto your redis server

So you suggest the Redis server can be compromised, but your web server can not?
If PHP source code is exposed - all encryption fails since encryption keys should be somehow stored at PHP program level.

**Comment by samdark** [comment]
Implementing such encryption by default is a no go. Recommending it for all cases is questionable as well because it affects performance significantly. If I store CC numbers or any really sensitive data, I'll encrypt them and only them.

**Comment by samdark** [comment]
@klimov-paul not necessary. Encryption key could be asked to be entered by user each time he accesses too sensitive data.

**Comment by cebe** [comment]
@klimov-paul noone said that is can not be broken, this is to reduce the risk. Breaking two servers at the same time is a bit less likely then one server.

**Comment by samdark** [comment]
But that's, again, very very application

**Comment by tom--** [comment]
@samdark right.

now because yii provides a high level api for caching and session, the crypto functions can be common to all the storage backends and transparnt to the user's api. hence i think the framework is a good place for it.

**Comment by klimov-paul** [comment]
> not necessary. Encryption key could be asked to be entered by user each time he accesses too sensitive data.

Each time before session saving?

**Comment by samdark** [comment]
> Each time before session saving?

That's why I'm talking about handling sensitive data in a special way and not doing it for session.

**Comment by cebe** [comment]
no of course not...

**Comment by klimov-paul** [comment]
> noone said that is can not be broken, this is to reduce the risk. Breaking two servers at the same time is a bit less likely then one server.

I can see no difference. hacking a single Redis server or signle web server makes no difference to me.
What make you think hacking Redis is more easier task then hacking web server?

**Comment by samdark** [comment]
Usually it's more complicated since a server with Redis only is much simpler.

**Comment by cebe** [comment]
@klimov-paul that is not the point of the discussion. Whether to encrypt data or not is up to decision by app developer and sysadmins. It should just be possible with yii.

**Comment by klimov-paul** [comment]
> If I store CC numbers or any really sensitive data, I'll encrypt them and only them.

This matter does not belongs to this topic. We have `Security` component, which allows you to encrypt any data, which is sensitive in the scope of your project at will. It is a matter of business logic - not the framework.

This issue demands some automatical encryption for the session or cache data.

**Comment by cebe** [comment]
> This issue demands some automatical encryption for the session or cache data.

It should not be done by default. It should be possible to do it if needed. Lets check whether Session and Cache component are extensible enough and if so, the issue is already done. If not, we can add some extension point.

**Comment by klimov-paul** [comment]
> that is not the point of the discussion. Whether to encrypt data or not is up to decision by app developer and sysadmins. It should just be possible with yii.

As I said Yii already provides `Security` component, which can be used for encryption at will.
I do not agree that framework should encrypt entire(!) session data. If you store sensitive information in the session and you afraid it can be compromised you should encrypt it manually according to your business logic.

**Comment by klimov-paul** [comment]
Assume we add encryption for the Cache data - it will encrypt any incoming message including such information as DB table schemas for example, reducing performance without a reason.

**Comment by cebe** [comment]
you can have a different cache component for that.

**Comment by cebe** [comment]
There is the following:

- As @samdark pointed out, there is `Cache::$serializer` which can do encryption for storing data in the cache.
- There is also `CacheSession` which can use the cache with serializer to encrypt the whole session if needed.
- For single values there is `Yii::$app->security->encryptBy*()` functions.

So all that is needed is some descirption in the guide on how to use these if needed imo.

**Comment by samdark** [comment]
Agree.

**Comment by samdark** [comment]
@tom-- ?

**Comment by tom--** [comment]
sorry. i was so disoriented by the previous conversation that i went to do other work for a while.

as far as cache goes, that sounds fine.

but we can't we also (optionally) add crypto layer when a session storage backend (db, redis, cache) attaches its handler?

**Comment by samdark** [comment]
Could you elaborate?

**Comment by cebe** [comment]
> but can't we also (optionally) add crypto layer when a session storage backend (db, redis, cache) attaches its handler?

isn't this the same as 

> There is also CacheSession which can use the cache with serializer to encrypt the whole session if needed.

Possible config:

```php
'components' => [
    'sessionStorage' => [
        'class' => yii\redis\Cache::class,
        'serializer' => [function($data) { /* encrypt here*/}, function($data) { /* dencrypt here*/}],
    ],
    'session' => [
        'class' => yii\web\CacheSession::class,
        'cache' => 'sessionStorage'
    ],
],
```

**Comment by tom--** [comment]
I have something a bit different in mind. But I need a little time to elaborate it to the point that I can explain it.

**Comment by tom--** [comment]
Perhaps roughly like [this](https://github.com/yiisoft/yii2/compare/master...tom--:crypto-session?expand=1)? Completely untested! Just trying to show the idea.

Then we would revise the session back-ends, such as redis or cache, to use `Session::$handler` rather than `getUseCustomStorage()`

**Comment by tom--** [comment]
the branching is still wrong in `registerSessionHandler()`

**Comment by tom--** [comment]
https://github.com/yiisoft/yii2/compare/master...tom--:crypto-session?expand=1 

Works with PHP's internal session handler using `./yii serve`. I haven't tested with other session handlers yet.

There's some other details to consider with the key. I would prefer to combine an app key with a user or session identifier, e.g. using session ID.

```php
        'session' => [
            'cryptoKey' => function ($id) {
                return Yii::$app->security->hkdf('SHA256', SESSION_CRYPTO_KEY, null, $id, 32);
            },
        ],
```

But this requires care when we change the session ID: we need to update `yii\web\Session` to read session before changing session ID and then write it again. Alternatively one could use in HKDF something not liable to change, e.g. user ID or a value from a cookie.

Other other work is to update the back-ends to use `yii\web\Session::$handler` instead of `yii\web\Session::$useCustomStorage`

Based on this experiment, it seems Yii could support session encryption in the framework. Any interest?

**Comment by cebe** [comment]
sounds interesting, has a low priority though.

**Comment by klimov-paul** [comment]
Note, that encryption of the session data is not enough - you'll need to encrypt session ID as well.
In case session ID is exposed attacker can use it accessing current user session and gaining access to the user personal data.

**Comment by tom--** [comment]
In the example above, the Session ID is encrypted by HKDF using the master session key. It's still not a very good key management scheme but it's probably an improvement over just using the master key. I prefer the idea of allocating a key per user and keeping it in a cookie.

What this highlights is that key management is often the hard part of a crypo system. This proposal isn't adequate if it requires the users to figure out that part. The framework would need to include that as part of a session crypto feature.

**Comment by rhertogh** [comment]
According to this post http://stackoverflow.com/questions/549/the-definitive-guide-to-form-based-website-authentication (and other articles), it is recommended that the session id is hashed before storing

> If you DO decide to implement persistent login cookies, this is how you do it:
> First, take some time to read Paragon Initiative's article on the subject. You'll need to get a bunch of elements right, and the article does a great job of explaining each.
> And just to reiterate one of the most common pitfalls, DO NOT STORE THE PERSISTENT LOGIN COOKIE (TOKEN) IN YOUR DATABASE, ONLY A HASH OF IT! The login token is Password Equivalent, so if an attacker got their hands on your database, they could use the tokens to log in to any account, just as if they were cleartext login-password combinations. Therefore, use hashing (according to http://security.stackexchange.com/a/63438/5002 a weak hash will do just fine for this purpose) when storing persistent login tokens. (SHA-2 is recommended)

Especially in the light of the recent Reddis DB leakages a hashed session id would make sense.

@tom-- When only the hashed session id is stored the session id could be combined with the system wide key to encrypt the data (salt & pepper)

In addition to that, shouldn't we implement salt & pepper (https://blog.filippo.io/salt-and-pepper/) for the password storage as well?

**Comment by cebe** [comment]
> In addition to that, shouldn't we implement salt & pepper (https://blog.filippo.io/salt-and-pepper/) for the password storage as well?

that deserves a separate issue, please open a new issue for that.

**Comment by spinitron** [comment]
@rhertogh this issue is about encryption of data stored in caches and in session backend. the topics you mention are both different.

@cebe already mentioned pepper.

On the matter of storing hashes of "remember me" cookie auth-keys, this is currently beyond the framework's scope. One would compare hashes in [yii\web\IdentityInterface::validateAuthKey()](http://www.yiiframework.com/doc-2.0/yii-web-identityinterface.html#validateAuthKey()-detail) which is up to the user.
